### ISSUE TITLE
Discrepancy between FilesModel::findById and Model::findById

### ISSUE THREAD

**Issue by fritzmg** [contao/contao#5033]
### Affected version(s)

5.0

### Description

In Contao 4.13 we deprecated being able to pass `null` to `Model::findById` and in `5.0` this is now throwing an exception. However the same does not apply to `FilesModel::findById`. Contao won't check for `null` there - but later on the following error will occur:

```
Doctrine\DBAL\Exception\SyntaxErrorException:
An exception occurred while executing a query: SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '? LIMIT 0,1' at line 1

  at vendor\doctrine\dbal\src\Driver\API\MySQL\ExceptionConverter.php:88
  at Doctrine\DBAL\Driver\API\MySQL\ExceptionConverter->convert(object(Exception), object(Query))
     (vendor\doctrine\dbal\src\Connection.php:1814)
  at Doctrine\DBAL\Connection->handleDriverException(object(Exception), object(Query))
     (vendor\doctrine\dbal\src\Connection.php:1749)
  at Doctrine\DBAL\Connection->convertExceptionDuringQuery(object(Exception), 'SELECT tl_files.* FROM tl_files WHERE tl_files.`id`=? LIMIT 0,1', array(), array())
     (vendor\doctrine\dbal\src\Connection.php:1055)
  at Doctrine\DBAL\Connection->executeQuery('SELECT tl_files.* FROM tl_files WHERE tl_files.`id`=? LIMIT 0,1', array(), array())
     (vendor\contao\contao\core-bundle\contao\library\Contao\Database\Statement.php:268)
  at Contao\Database\Statement->query('', array())
     (vendor\contao\contao\core-bundle\contao\library\Contao\Database\Statement.php:222)
  at Contao\Database\Statement->execute()
     (vendor\contao\contao\core-bundle\contao\library\Contao\Model.php:1154)
  at Contao\Model::find(array('limit' => 1, 'column' => 'id', 'value' => array(), 'return' => 'Model', 'table' => 'tl_files', 'offset' => 0))
     (vendor\contao\contao\core-bundle\contao\library\Contao\Model.php:982)
  at Contao\Model::findOneBy('id', null, array('limit' => 1, 'column' => 'id', 'value' => null, 'return' => 'Model'))
     (vendor\contao\contao\core-bundle\contao\models\FilesModel.php:136)
  at Contao\FilesModel::findById(null)
```

Was this an oversight and we need to add a null check to `FilesModel::findById`? Or do we deliberately allow `null` there and we need to fix the subsequent error?

**Comment by ausi** [comment]
> Was this an oversight and we need to add a null check to `FilesModel::findById`?

IMO yes.

**Comment by fritzmg** [comment]
The issue is rather about `Model::findBy` vs `Model::findOneBy`. The latter does not throw an exception for the primary key currently when passing `null` as the value.

**Comment by ausi** [comment]
So `Model::findById(null)` throws while `Model::findOneById(null)` does not?
If so, we should fix `Model::findOneBy…()` IMO.

**Comment by fritzmg** [comment]
Yes. However, why did we deprecate passing `null` in the first place?  Seems odd that `Model::findById(null)` throws an exception but `Model::findByFoobar(null)` (where `foobar` is not nullable) doesn't.

**Comment by ausi** [comment]
> Yes. However, why did we deprecate passing `null` in the first place?

Maybe related to the changes from #3207 ?

**Comment by fritzmg** [comment]
Yes, the deprecation was added there - but I am not sure why 🤔

**Comment by fritzmg** [comment]
Just to give an example how I stumbled over this: in fritzmg/contao-file-access I was traversing up the directory structure this way:

```php
$filesModel = FilesModel::findOneByPath($path);

do {
    // …
    $filesModel = FilesModel::findById($filesModel->pid);
} while (null !== $filesModel);
```

`$filesModel->pid` will at some point be `null` for the top most folder in the hierarchy and thus this exception occurs. For Contao 5.0 I have to use

```php
$filesModel = FilesModel::findOneByPath($path);

do {
    // …
    $filesModel = $filesModel->pid ? FilesModel::findById($filesModel->pid) : null;
} while (null !== $filesModel);
```

now.

**Comment by ausi** [comment]
So, should we just return `null` instead of an exception in all these cases?

**Comment by fritzmg** [comment]
I think so, as foreign keys are potentially nullable and doctrine also does not care if you execute something like

```php
$db->fetchAssociative("SELECT * FROM tl_page WHERE id = ?", [null])
```

**Comment by m-vo** [comment]
While primary keys are never nullable, the operation `findBy*` with a nullable FK is essential. And I don't see a reason why we should make a difference in the API (PK vs FK). I would also vote for handling that case and just return `null` like before.

**Comment by fritzmg** [comment]
Also the following error currently occurs in the back end when clicking _Save & New_ when editing a content element for example:

```
TypeError:
Model::findByPk(): Argument #1 ($varValue) must be of type int|string, got null

  at vendor\contao\contao\core-bundle\contao\library\Contao\Model.php:820
  at Contao\Model::findByPk(null)
     (vendor\contao\contao\core-bundle\contao\dca\tl_content.php:981)
  at tl_content->adjustDcaByType(object(DC_Table))
     (vendor\contao\contao\core-bundle\contao\drivers\DC_Table.php:206)
  at Contao\DC_Table->__construct('tl_content', array('tables' => array('tl_article', 'tl_content'), 'table' => array('Contao\\CoreBundle\\Controller\\BackendCsvImportController', 'importTableWizardAction'), 'list' => array('Contao\\CoreBundle\\Controller\\BackendCsvImportController', 'importListWizardAction')))
     (vendor\contao\contao\core-bundle\contao\classes\Backend.php:293)
  at Contao\Backend->getBackendModule('article', null)
```

**Comment by ausi** [comment]
> Also the following error currently occurs in the back end when clicking _Save & New_ when editing a content element

Just stumbled myself over the same issue.
Do you plan on making a PR that allows `null` for those methods?

**Comment by ausi** [comment]
@fritzmg this can be closed now since #5097 was merged, right?

**Comment by fritzmg** [comment]
Yes :)
